os_stub/memlib/copy_mem.c

```diff
@@ -9,35 +9,65 @@
  **/
 
 #include "base.h"
+#include "library/debuglib.h"
 
 /**
- * Copies a source buffer to a destination buffer, and returns the destination buffer.
+ * Copies bytes from a source buffer to a destination buffer.
  *
- * This function copies length bytes from source_buffer to destination_buffer, and returns
- * destination_buffer.  The implementation must be reentrant, and it must handle the case
- * where source_buffer overlaps destination_buffer.
+ * This function copies "len" bytes from "src_buf" to "dst_buf".
  *
- * If length is greater than (MAX_ADDRESS - destination_buffer + 1), then ASSERT().
- * If length is greater than (MAX_ADDRESS - source_buffer + 1), then ASSERT().
+ * Asserts and returns a non-zero value if any of the following are true:
+ *   1) ("src_buf" or "dst_buf" are NULL) and "len" is greater 0.
+ *   2) "src_buf" and "dst_buf" overlap.
+ *   3) "len" is greater than "dst_len".
+ *   4) "len" is greater than (MAX_ADDRESS - "dst_buf" + 1).
+ *   5) "len" is greater than (MAX_ADDRESS - "src_buf" + 1).
+ *   6) "len" is greater than (SIZE_MAX >> 1).
  *
```

Comment:
I think we need add 4~6) shall also be applicable to "dst_len".

Reply:
I think 3 implies 4-6 for "dst_len", but I can call it out explicitly.

Reply:
Ok.

Reply:
I think 4-6 are applicable for "dst_len", but the additional checks (4-6) against "dst_len" are unnecessary due to 3. Can someone else confirm?

Reply:
3) is not enough.

For example: len == 10, dst_len == 50, (size_max >> 1) == 20.

It can pass all checks above. But it is invalid, because dst_len is too large. There must be something wrong.

Reply:
Ok. Thanks.